### PR TITLE
Implement ContextStorageOverride for opentelemetry context bridge

### PR DESCRIPTION
**Pull request by YifeiZhuang**
b/357892637
follow up with https://github.com/grpc/grpc-java/pull/11500

opencensus shim will find opentelmetry context when using opencensus {@link io.opencensus.trace.unsafe.ContextManagerImpl}, but does not solve the problem of bridging io.grpc.context and io.opentelemetry.context

otel implementation does not have otel over grpc contxt: 
https://github.com/open-telemetry/opentelemetry-java/blob/18d192d1fadc6f6dd4dfbff01c0c139eaf93c604/context/src/grpcInOtelTest/java/io/grpc/override/ContextStorageOverride.java

other implementation: 
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/grpc-1.6/library#library-instrumentation-for-grpc-160

### CODE REVIEW

File: contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    return previous.withValue(OTEL_SCOPE, scope);
+  }
+
+  @Override
+  public void detach(Context toDetach, Context toRestore) {
+    Scope scope = OTEL_SCOPE.get(toRestore);
+    if (scope == null) {
+      logger.log(
+          Level.SEVERE, "Detaching context which was not attached.",
+          new Throwable().fillInStackTrace());
```
Comment (ejona86) [inline contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java:58]: Is `fillInStackTrace()` necessary?

File: contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    } else {
+      scope.close();
+    }
+  }
+
+  @Override
+  public Context current() {
+    io.opentelemetry.context.Context otelCurrent = io.opentelemetry.context.Context.current();
+    io.grpc.Context grpcCurrent = otelCurrent.get(GRPC_CONTEXT_OVER_OTEL);
+    if (grpcCurrent == null) {
```
Comment (ejona86) [inline contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java:68]: FYI: you could use `Context.keyWithDefault("blah", Context.Root)` to let the API handle this for you.

File: settings.gradle [hunk 3]
```diff
@@ -76,6 +76,7 @@ include ":grpc-istio-interop-testing"
 include ":grpc-inprocess"
 include ":grpc-util"
 include ":grpc-opentelemetry"
+include ":grpc-context-storage-override"
```
Comment (ejona86) [inline settings.gradle:79]: The name of the package is clearly io.grpc.context-related. However, nothing speaks to "otel" in the name. `context-opentelementry-storage` (+override?) as the name? (`context-storage-override-opentelemetry`?)

File: contextstorage/build.gradle [hunk 4]
```diff
@@ -0,0 +1,27 @@
+plugins {
+    id "java-library"
+    id "maven-publish"
```
Comment (ejona86) [inline contextstorage/build.gradle:3]: Let's remove maven-publish until we are confident we like the name. With this removed, we don't have to worry about accidentally publishing it.

**Review by ejona86 — APPROVED** [review]
Only concern is the name.

File: contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java (hunk 2 above)
Comment (YifeiZhuang) [reply contextstorage/src/main/java/io/grpc/override/OpenTelemetryContextStorage.java:68]: hm, this only works for `io.opentelemetry.context.keyWithDefault` but they don't have this convenient API.

File: settings.gradle [hunk 5]
```diff
@@ -111,6 +112,7 @@ project(':grpc-istio-interop-testing').projectDir = "$rootDir/istio-interop-test
 project(':grpc-inprocess').projectDir = "$rootDir/inprocess" as File
 project(':grpc-util').projectDir = "$rootDir/util" as File
 project(':grpc-opentelemetry').projectDir = "$rootDir/opentelemetry" as File
+project(':grpc-opentelemetry-context-storage-override').projectDir = "$rootDir/contextstorage" as File
```
Comment (YifeiZhuang) [inline settings.gradle:115]: API discussion:
`grpc-opentelemetry-context-override`

[x]`grpc-context-override-opentelemetry`

`grpc-context-opentelemetry-override`

**Pull request merged**